New ui dave a029

juzen2003 opened the pull request:
# This is a dummy pull request to track the update on MutationObserver
# Update:
* A029:
    * search page:
        * Add mutationObserver to detect changes in search sidebar and update ps
            * Test cases:
                1. Change browser size in other tabs and switch to search, ps will get updated
                2. If left pane is just large enough to hold the contents without scrollbar, by clicking any dropdown arrow to expand a new menu, ps will show up (get updated). Undo previous action, ps will disappear.
                3. If left pane is just large enough to hold the contents without scrollbar, by clicking any input in widgets that adds a new item in the left pane, ps will show up (get updated). Undo previous action, ps will disappear.
                4. collapse/expand widgets, ps will get updated.
        * Add mutationObserver to detect changes in search widgets and update ps
            * Test cases:
                1. Add widgets until the last one is just off the bottom of the browser, ps will show up
                2. Change browser size in other tabs and switch to search, ps will get updated  
                3. Add/Remove/collapse/expand widgets, ps will get updated.

# Note:
* For now only two ps in search page are moved to MutationObserver.
* Currently update calls for those two ps are all at line 453-535 in opus.js. Will have to create a function for all codes related to MutationObserver eventually  
* All adjustSearchSideBarHeight, adjustSearchWidgetHeight, adjustSearchHeight scattered in other files are commented out.
* Reference: 
  * https://www.javascripture.com/MutationObserver
  * https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/MutationObserver
  * Online code editor for experiment:
    * http://cssdeck.com/labs/badnvznx

juzen2003 committed:
Setup MutationObserver for experiment in opus.js

juzen2003 committed:
Move adjustSearchSideBarHeight and adjustSearchWidgetHeight to the callback function of MutationObserver

juzen2003 committed:
Update MutationObserver to watch for changes in search sidebar and widgets, and perform ps update when needed (search page)

juzen2003 committed:
Add MutationObserver to watch for switching tabs (class changes) and update ps correspondingly

juzen2003 committed:
Rename switchTabObserver to adjustAllPSObserver

juzen2003 committed:
Update callback functions in searchWidgetObserver & searchSidebarObserver

juzen2003 committed:
Add mutationObserver to cart page

juzen2003 committed:
Modify mutationObserver in cart page

juzen2003 committed:
- Remove mutationObserver in detail page because the ps is initialized when all info are returned from multiple ajax calls, and every time we switch to detail page, getDetail will get called and render the html again (re-init ps as well).

juzen2003 committed:
Add mutationObserver to help panel

juzen2003 committed:
Remove unused id in help panel

juzen2003 committed:
Update mutationObserver in helpPanel

juzen2003 committed:
Add mutationObserver for select metadata modal search menu

juzen2003 committed:
Add mutationObserver to selected metadata area in the modal

juzen2003 committed:
Add mutationObserver to browse dialog

juzen2003 committed:
Add mutationObserver to browse gallery and table views

juzen2003 committed:
Clean up codes

juzen2003 committed:
Merge with upstream/new_ui

juzen2003 committed:
Clean up codes

juzen2003 committed:
Fix S001

dstopp reviewed (changes requested):
- i think we can remove the commented out code
- try using $( "#foo" )[ 0 ]; to get DOM element per conversation to avoid adding all the id's
- move observer code to its own namespace
- opus.js - I prefer switch to if then else then if because it feels easier to read.

juzen2003 committed:
Fix S031

juzen2003 committed:
Update all ps to have minScrollbarLength set to 30

juzen2003 committed:
Fix code style in observePerfectScrollbar function

juzen2003 committed:
Commit before switching to another branch

juzen2003 committed:
Switch all getElementById(selector) to [0]

juzen2003 committed:
Move mutationObserver to another namespace

juzen2003 committed:
- Clean up all commented out ps update
- Remove unwanted logic block of load function in opus.js

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -390,7 +386,7 @@ var opus = {
             opus.helpScrollbar.update();
         }
-    }
+    },
```
Does this comma serve a purpose?

juzen2003 replied:
It's removed now. It was there because there was another function down there in the same namespace.

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -313,11 +312,8 @@ var opus = {
-            let adjustSearchWidgetHeight = _.debounce(o_search.adjustSearchHeight, 800);
-            adjustSearchWidgetHeight();
-        });
+        // wait until all widgets are get, do somthing here:
```
Can we just get rid of these lines?

juzen2003 replied:
Commented out codes are removed.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1440,6 +1435,54 @@ var o_browse = {
         o_browse.tableScrollbar.update();
     },
 
+    adjustmetadataModalMenu: function() {
```
What is a metadataModalMenu? It's not obvious to me what menu we're talking about.

juzen2003 replied:
Rename it to adjustMetadataSelectorMenu

juzen2003 committed:
Clean up codes

dstopp reviewed (changes requested):
- i think we can remove the commented out code
- try using $( "#foo" )[ 0 ]; to get DOM element per conversation to avoid adding all the id's
- move observer code to its own namespace
- opus.js - I prefer switch to if then else then if because it feels easier to read.

-in the instantiation of the scrollbars, are tableScrollbar & galleryScrollbar created to be generic for both browse and cart?

-do we want to set the minScrollbarLength of 30 as a global const and use it throughout the file instead of '30'? <probably should live in opus.js since it's used all over>

-how does the type of functionality of 'adjustMetadataSelectorMenu' diff from 'adjustSelectedMetadataPS' ?  if they are both adjusting the scrollbar, might want to name similar for consistency.

-detail.js - why is the the scrollbar initialized everytime?  can't we init it like we do in browse and cart just once?

-don't we need to specify the namespace <browse vs cart> for the scrollbar?  

-in opus.js, why did you remove the if ($.isEmptyObject..) at line 112?
-in opus.js, if you don't change the lines at 122-126, we will avoid a conflict when we merge <changes are not necessary>

juzen2003 committed:
- opus.minimumPSLength is created (set to 30) in opus.js
- Rename "adjustMetadataSelectorMenu" to "adjustMetadataSelectorMenuPS"

juzen2003 committed:
- rename opus_mutationObserver to o_mutationObserver
- Add #browse to be more specific when initializing tableScrollbar and galleryScrollbar in browse page
- Add comments to initAndUpdatePerfectScrollbar in detail.js
- Update comments in mutationObserver for cart tab

juzen2003 committed:
Clean up codes

juzen2003 committed:
In mutationObserver.js, removed unused target node: detailTab = [0]
